enhancements/network/allow-mtu-and-overlay-port-changes.md

```diff
@@ -0,0 +1,165 @@
+tunnels post-installation. However these changes aren't trivial and may cause
+downtime, hence the CNO forbids currently them.
+
+We propose a brand new daemonset that will run on every node of the cluster and
+will make the necessary changes in an ordered, and coordinated manner with a
+service disruption within the TCP timeout.
+
+## Motivation
+
+While customers usually set the MTU, and the overlay ports some times they
+need to change them afterwards for unpredictable reasons such as performance,
+or changes in the underlay.
+
+## Goals
+
+* Allow to change MTU post install on both OpenShift SDN and OVN Kubernetes.
+
+* Allow to change the overlay network ports on the underlay in both OpenShift
+  SDN and OVN Kubernetes.
+
+* Allow to change both vxlan and geneve ports in a single operation.
+
+## Non goals
+
+* Allow to change MTU configuration in te underlay interfaces.
+
+* Allow to change MTU, and either port in a single operation.
+
+* Other safe or unsafe configuration changes.
+
+## Proposal
+
+When the cluster network operator detects the change it will deploy a new
+deaemonset with restartPolicy Mever which is responsible for doing the changes.
+
+Even though both changing the MTU, and changing the port have a lot in common,
+there are some key differences and therefore the daemonset will be slightly
+different.
+
+When decreasing the MTU each of the deamonset will:
+
+1. Verify that the new MTU is valid for that node. If it is valid add to
+   itself the annotation `openshift.io/cno-able-to-change-mtu`.
+2. Wait until the timestamp defined in `CHANGE_MTU_INIT_TS`
+3. Verify that every pod of that daemonset is running and has the annotation
+   `openshift.io/cno-able-to-change-mtu` if any of the pods is not running, or
+   is missing that annotation the pod will abort the procedure.
+4. Cordon the node where its running.
+5. Wait a grace period. This grace period is based on
+   `CHANGE_MTU_INIT_TS` rather than the time of finalization of the
+   previous step.
+6. Proceed to change the MTU of `eth0` on each one of the pods netnamespace.
+7. Wait a grace period. This grace period is based on
+   `CHANGE_MTU_INIT_TS` rather than the time of finalization of the
+   previous step.
+8. Proceed to change the MTU of each pod in the virtual switch.
+9. Wait a grace period. This grace period is based on
+   `CHANGE_MTU_INIT_TS` rather than the time of finalization of the
+   previous step.
+10. In the case of OVN Kubernetes proceed to change the MTU of the interfaces
+   ovn-k8s-mp0, br-local, and ovn-k8s-gw0 sequentially without any delay in
+   between.
+
+   In the case of OpenShiftSDN proceed to change the MTU of the interfaces
+   vxlansys.
+
+13. Uncordon the node
+
+14. Annotate itself with `openshift.io/cno-succesuflly-changed-mtu` and exit
+    with code 0.
+
+When increasing the MTU the same process applies, except steps 10, 8, and 6 are done in reverse order.
+
```

Comment:
When Increasing we should be able to just set the MTU values larger.  Any existing TCP (or SCTP) connection will not be able to make use of the larger MSS at all.  The active MSS used  has long ago been negotiated when the 3WHS established the TCP connection, or been since set via an ICMP triggered along the path.   Likewise for UDP.  Either the new larger MTU is used for the next transmit, or the old one (e.g. if cached) is used, and will still work.

Reply:
existing connections won't notice the change, but new connections could; if we raised the pod MTU first, then a pod might try to open a new connection using the new MTU and then smash into the lower MTU on the bridge